ledger/participant-integration-api/src/main/scala/platform/store/dao/events/TransactionsReader.scala

```diff
@@ -98,41 +108,217 @@ private[dao] final class TransactionsReader(
+            queryNonPruned.executeSql(
+              query = eventStorageBackend.fetchFlatConsumingEvents(
+                eventSequentialIds = ids,
+                allFilterParties = allFilterParties,
+              )(connection),
+              minOffsetExclusive = firstOffset,
+              error = (prunedOffset: Offset) =>
+                s"Transactions request from ${firstOffset.toHexString} to ${lastOffset.toHexString} precedes pruned offset ${prunedOffset.toHexString}",
+            )
+          }
+        )
+    }
+
+    def fetchCreateEvents(
+        ids: Iterable[Long]
+    ): Future[Vector[EventStorageBackend.Entry[Raw.FlatEvent]]] = {
+      eventFetchingLimiter
+        .execute(
+          dispatcher.executeSql(metrics.daml.index.db.getFlatTransactions) { implicit connection =>
+            queryNonPruned.executeSql(
+              query = eventStorageBackend.fetchFlatCreateEvents(
+                eventSequentialIds = ids,
+                allFilterParties = allFilterParties,
+              )(connection),
+              minOffsetExclusive = firstOffset,
+              error = (prunedOffset: Offset) =>
+                s"Transactions request from ${firstOffset.toHexString} to ${lastOffset.toHexString} precedes pruned offset ${prunedOffset.toHexString}",
+            )
+          }
+        )
+    }
+
+    import scala.util.chaining._
+
+    val consumingStream: Source[EventStorageBackend.Entry[Raw.FlatEvent], NotUsed] = filters
+      .map(streamConsumingIds_stakeholders)
+      .pipe(FilterTableACSReader.mergeSort[Long])
+      .statefulMapConcat(statefulDeduplicate)
+      .via(
+        BatchN(
+          maxBatchSize = pageSize,
+          maxBatchCount = consumingEventsFetchingParallelism_mapAsync + 1,
+        )
+      )
+      .async
+      .addAttributes(
+        Attributes.inputBuffer(
+          initial = consumingEventsFetchingParallelism_mapAsync,
+          max = consumingEventsFetchingParallelism_mapAsync,
+        )
+      )
+      .mapAsync(consumingEventsFetchingParallelism_mapAsync)(fetchConsumingEvents)
+      .mapConcat(identity)
+
+    val createStream: Source[EventStorageBackend.Entry[Raw.FlatEvent], NotUsed] = filters
+      .map(streamCreateIds_stakeholders)
+      .pipe(FilterTableACSReader.mergeSort[Long])
+      .statefulMapConcat(statefulDeduplicate)
+      .via(
+        BatchN(
+          maxBatchSize = pageSize,
+          // TODO pbatko: Why +1 ?
+          maxBatchCount = createEventsFetchingParallelism_mapAsync + 1,
+        )
+      )
+      .addAttributes(
+        Attributes.inputBuffer(
+          initial = createEventsFetchingParallelism_mapAsync,
+          max = createEventsFetchingParallelism_mapAsync,
+        )
+      )
+      .mapAsync(createEventsFetchingParallelism_mapAsync)(fetchCreateEvents)
+      .mapConcat(identity)
+
+    val allFlatEvents: Source[EventStorageBackend.Entry[Raw.FlatEvent], NotUsed] =
+      consumingStream.mergeSorted(createStream)(
```

Comment:
something like this would work too: Ordering.by[EventStorageBackend.Entry[Raw.FlatEvent], Long](_.eventSequentialId)

Reply:
Ok